File: data/conala-train_code-davinci-002_1666524963/Java/34.java
public static <T> List<List<T>> allCombinations(List<List<T>> lists) {
    List<List<T>> result = new ArrayList<>();
    allCombinations(lists, 0, new ArrayList<>(), result);
    return result;
}

private static <T> void allCombinations(List<List<T>> lists, int depth, List<T> current, List<List<T>> result) {
    if (depth == lists.size()) {
        result.add(current);
        return;
    }
    for (T element : lists.get(depth)) {
        List<T> copy = new ArrayList<>(current);
        copy.add(element);
        allCombinations(lists, depth + 1, copy, result);
    }
}

File: data/conala-train_code-davinci-002_1666524963/Python/34.py
def all_combinations(lists):
    """
    Returns all combinations of a list of lists.
    """
    if len(lists) == 0:
        return []
    elif len(lists) == 1:
        return lists[0]
    else:
        return [x + [y] for x in all_combinations(lists[:-1]) for y in lists[-1]]